Statistics : add info related to the study

flomnes opened the pull request:
TODO
- [x] Fill the `StudyInfo` structure from the study using the `initializeFromStudy` method, probably just after the study has been loaded
- [x] Write the data into a separate file, named `study-info.txt` in the output directory

flomnes committed:
Rename TimeElapsed -> Benchmarking

flomnes committed:
Rename timeelapsed -> benchmarking

flomnes committed:
Rename info.h -> timeinfo.h

flomnes committed:
Add struct StudyInfo

guilpier-code committed:
study info file for statistics : dump study number of years and number of enabled thermal clusters

flomnes commented on src/libs/antares/benchmarking/studyinfo.cpp:
```diff
@@ -1,12 +1,53 @@
+			auto end = area.thermal.list.end();
+			for (auto i = area.thermal.list.begin(); i != end; ++i)
+			{
+				ThermalCluster* cluster = i->second.get();
+				if (cluster->enabled)
+					nbEnabledThermalClusters_++;
```
No need to get the raw pointer to de-reference, `std::unique_ptr<T>` defines a structure dereference operator (`->`)
```suggestion
				if (i->second->enabled)
					nbEnabledThermalClusters_++;
```

guilpier-code replied:
ok, but we could leave the code as is for more clarity : 
```C++
if (cluster->enabled)
```
Is quite clear I guess

flomnes commented on src/libs/antares/benchmarking/studyinfo.h:
```diff
@@ -1,33 +1,40 @@
+private:
+    // Methods
+    void getNbPerformedYears();
+    void getEnabledThermalClusters();
```
I suggest passing `study` to each method, not to the class constructor.
```suggestion
    StudyInfo();
    void collect(const Antares::Data::Study& study);
    void flush(Yuni::String& filePath);

private:
    // Methods
    void getNbPerformedYears(const Antares::Data::Study& study);
    void getEnabledThermalClusters(const Antares::Data::Study& study);
```

flomnes commented on src/solver/application.cpp:
```diff
@@ -234,6 +235,16 @@ void Application::prepare(int argc, char* argv[])
+    study_info.collect();
+
+    // Flush previous info into a record file
+    Yuni::String filePath; 
+    filePath.clear() << pStudy->folderOutput << Yuni::IO::Separator << "study-info.txt";;
```
You may use pStudy->buffer

See **study.h**
```cpp
class Study final : public Yuni::NonCopyable<Study>, public IObject
{
// ...
mutable YString buffer;
// ...
};
```

flomnes commented on src/solver/application.cpp:
```diff
@@ -234,6 +235,16 @@ void Application::prepare(int argc, char* argv[])
+    // Flush previous info into a record file
+    Yuni::String filePath; 
+    filePath.clear() << pStudy->folderOutput << Yuni::IO::Separator << "study-info.txt";;
+    study_info.flush(filePath);
```
I suggest splitting responsabilities :

- 1 class stores data
- Multiple classes can write data in CSV, human-readable format, JSON, etc.
- Possibly one interface class for those writers

flomnes commented on src/libs/antares/benchmarking/studyinfo.cpp:
```diff
@@ -1,12 +1,53 @@
+			throw Antares::Error::CreatingStudyInfoFile(filePath);
+		}
+
+		outputFile_ << "Number of performed years" << "\t" << nbPerformedYears_ << "\n";
+		outputFile_ << "Number of enabled thermal clusters" << "\t" << nbEnabledThermalClusters_ << "\n";
```
Since this data is meant to be parsed for statistical analysis, I suggest making it less human-readable, and more easily parsable. Avoid spaces, use short names, etc.

guilpier-code replied:
I agree. Could you be more specific about this file format ?

flomnes replied:
```
areas <number of areas>
years <number of years>
thermal_clusters <number of thermal clusters>
etc.
```

flomnes commented on src/libs/antares/benchmarking/studyinfo.h:
```diff
@@ -1,33 +1,40 @@
+    unsigned int nbEnabledBindingConstraints_ = 0;
+    unsigned int nbAreas_ = 0;
+    unsigned int nbLinks_ = 0;
+    unsigned int UnitComitmentMode_;
```
```suggestion
    unsigned int UnitComitmentMode_ = 0;
```

flomnes commented on src/libs/antares/benchmarking/studyinfo.h:
```diff
@@ -1,33 +1,40 @@
-    unsigned int nbCoreMode;
+    unsigned int maxNbYearsInParallel_ = 0;
+    unsigned int minNbYearsInParallel_ = 0;
+    unsigned int nbCoreMode_;
```
```suggestion
    unsigned int nbCoreMode_ = 0;
```

guilpier-code committed:
study info file for statistics : dump areas and links count into study-info.txt

guilpier-code committed:
study info file for statistics : dump BC count, UC mode and nb of years in parallel into study-info.txt

guilpier-code committed:
study info file for statistics : separating file content and file writer into 2 different classes

guilpier-code committed:
study info file for statistics : file content class now has a vector of information item

Class StudyInfoContainer now has a vector of StudyInfoItem instead of a data member for each piece of information collected from study.
Class StudyInfoContainer still collects information on study but use the "addItem" simple method to add an information to its vector of information items.
The file writer now merely loops on file container's vector to print each info item.

guilpier-code committed:
study info file for statistics : the file's CSV writer now inherit form a base abstract class

StudyInfoWriter is now a base abstract class and its child StudyInfoCSVwriter is the writer to the CSV file.
Note that StudyInfoWriter is not exactly an interface, as it has data members supposed to be useful in any of its child writer,
especially a file content and a path to a file.

guilpier-code committed:
study info file for statistics : base writer should contain the content to be stored, but nothing else.

Therefore, our CSV writer now contains the file path as a data member.

flomnes reviewed (commented):
- Since we don't manipulate numerical data, you could just convert all data to std::string before adding it to the container (no need for a mixed-type container, nice interface though !)
- The data related to the optimization problem is missing (number of variables, constraints, non-zero terms in the constraint matrix, etc.), I'll try to add it.

flomnes committed:
Merge branch develop into feature/statistics-add-study-info

flomnes committed:
Create & call method Application::writeStudyInfos

flomnes committed:
Save optimization-related infos

flomnes committed:
Collect solver version

flomnes commented on src/libs/antares/benchmarking/studyinfo.h:
```diff
@@ -0,0 +1,130 @@
+    template<class ClassItem, typename T>
+    void addItem(std::string name, T value)
+    {
+        items_.push_back(std::unique_ptr<ClassItem>(new ClassItem(name, value)));
```
```suggestion
        items_.emplace_back(new ClassItem(name, value));
```

flomnes committed:
Update src/libs/antares/benchmarking/studyinfo.h

flomnes commented on src/libs/antares/benchmarking/studyinfo.cpp:
```diff
@@ -0,0 +1,161 @@
+			{
+				ThermalCluster* cluster = i->second.get();
+				if (cluster->enabled)
+					nbEnabledThermalClusters++;
```
```suggestion
				auto& cluster = i->second;
				if (cluster->enabled)
					nbEnabledThermalClusters++;
```

flomnes committed:
Apply suggestions from code review
